File: app/models/project.rb
class Project < ApplicationRecord
  CATEGORIES = %w[apps infrastructure developer-tools reference].freeze
  STATUSES   = %w[active shipped phase-0 legacy].freeze

  validates :slug, :name, :tagline, :category, :status, presence: true
  validates :slug, uniqueness: true

  scope :ordered,  -> { order(:position, :name) }
  scope :featured, -> { where(featured: true) }

  def stack_list
    stack.to_s.split(",").map(&:strip).reject(&:empty?)
  end

  def public_repo?
    visibility.to_s == "public" && repo.present?
  end
end

File: src/main/java/com/dunamismax/site/content/Project.java
package com.dunamismax.site.content;

import java.util.List;

public record Project(
    String slug,
    String name,
    ProjectCategory category,
    ProjectStatus status,
    int position,
    boolean featured,
    String visibility,
    String tagline,
    List<String> stack,
    String repo,
    String url) {

  public Project {
    stack = List.copyOf(stack);
  }

  public boolean publicRepo() {
    return "public".equals(visibility) && repo != null && !repo.isBlank();
  }

  public boolean isPublicRepo() {
    return publicRepo();
  }
}
